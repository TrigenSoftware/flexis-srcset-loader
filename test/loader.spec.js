import {
	compile,
	multiCompile,
	fs,
	pathToArtifacts
} from './compile';

jest.setTimeout(50000);

function findModuleSourceByName(stats, name) {
	if (!stats.modules) {
		return null;
	}

	for (const module of stats.modules) {
		if (module.name === name) {
			return module.source;
		}
	}

	for (const module of stats.modules) {
		const foundModuleSource = findModuleSourceByName(module, name);

		if (foundModuleSource) {
			return foundModuleSource;
		}
	}

	return null;
}

describe('srcset-loader', () => {
	it('should emit files', async () => {
		const stats = await compile('image.js', {
			rules: [{
				match: '(max-width: 3000px)',
				width: [1, 1920, 1280, 720, 560, 320],
				format: ['webp', 'jpg']
			}],
			scalingUp: false
		});
		const source = findModuleSourceByName(stats, './Felix.jpg');
		const commonjsSource = findModuleSourceByName(stats, './Felix.jpg?commonjs');
		const artifacts = fs.readdirSync(pathToArtifacts);

		expect(source.replace(
			/(__webpack_public_path__ \+ ").*(\.\w+")/gi,
			'$1asset$2'
		)).toMatchSnapshot();
		expect(commonjsSource.replace(
			/(__webpack_public_path__ \+ ").*(\.\w+")/gi,
			'$1asset$2'
		)).toMatchSnapshot();
		expect(artifacts.length).toBe(8);
	});

	it('should use cache with runtimeModulesCache option', async () => {
		// Warm up
		await multiCompile('image.js', {
			rules: [{
				match: '(max-width: 3000px)',
				width: [1, 1920, 1280, 720, 560, 320],
				format: ['webp', 'jpg']
			}],
			scalingUp: false,
			processOnce: false
		});
		await multiCompile('image.js', {
			rules: [{
				match: '(max-width: 3000px)',
				width: [1, 1920, 1280, 720, 560, 320],
				format: ['webp', 'jpg']
			}],
			scalingUp: false,
			processOnce: true
		});

		let stats = null;

		stats = await multiCompile('image.js', {
			rules: [{
				match: '(max-width: 3000px)',
				width: [1, 1920, 1280, 720, 560, 320],
				format: ['webp', 'jpg']
			}],
			scalingUp: false,
			processOnce: true
		});

		const timeWithCache = stats.children.reduce((sum, stats) => sum + stats.time, 0);

		stats = await multiCompile('image.js', {
			rules: [{
				match: '(max-width: 3000px)',
				width: [1, 1920, 1280, 720, 560, 320],
				format: ['webp', 'jpg']
			}],
			scalingUp: false,
			processOnce: false
		});

		const timeWithoutCache = stats.children.reduce((sum, stats) => sum + stats.time, 0);

		expect(timeWithoutCache).toBeGreaterThan(timeWithCache);
	});
});
