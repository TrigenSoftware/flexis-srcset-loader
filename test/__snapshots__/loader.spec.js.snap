// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`srcset-loader should emit files 1`] = `
"
export default __webpack_public_path__ + \\"asset.jpg\\";

export var srcset = [{
	format: \\"webp\\",
	type: \\"image/webp\\",
	name: \\"webp3000\\",
	width: 3000,
	height: 1500,
	src: __webpack_public_path__ + \\"asset.webp\\"
}, {
	format: \\"webp\\",
	type: \\"image/webp\\",
	name: \\"webp3200\\",
	width: 3000,
	height: 1500,
	src: __webpack_public_path__ + \\"asset.webp\\"
}, {
	format: \\"webp\\",
	type: \\"image/webp\\",
	name: \\"webp1920\\",
	width: 1920,
	height: 960,
	src: __webpack_public_path__ + \\"asset.webp\\"
}, {
	format: \\"webp\\",
	type: \\"image/webp\\",
	name: \\"webp1280\\",
	width: 1280,
	height: 640,
	src: __webpack_public_path__ + \\"asset.webp\\"
}, {
	format: \\"webp\\",
	type: \\"image/webp\\",
	name: \\"webp720\\",
	width: 720,
	height: 360,
	src: __webpack_public_path__ + \\"asset.webp\\"
}, {
	format: \\"webp\\",
	type: \\"image/webp\\",
	name: \\"webp560\\",
	width: 560,
	height: 280,
	src: __webpack_public_path__ + \\"asset.webp\\"
}, {
	format: \\"webp\\",
	type: \\"image/webp\\",
	name: \\"webp320\\",
	width: 320,
	height: 160,
	src: __webpack_public_path__ + \\"asset.webp\\"
}, {
	format: \\"jpg\\",
	type: \\"image/jpeg\\",
	name: \\"jpg3000\\",
	width: 3000,
	height: 1500,
	src: __webpack_public_path__ + \\"asset.jpg\\"
}, {
	format: \\"jpg\\",
	type: \\"image/jpeg\\",
	name: \\"jpg3200\\",
	width: 3000,
	height: 1500,
	src: __webpack_public_path__ + \\"asset.jpg\\"
}, {
	format: \\"jpg\\",
	type: \\"image/jpeg\\",
	name: \\"jpg1920\\",
	width: 1920,
	height: 960,
	src: __webpack_public_path__ + \\"asset.jpg\\"
}, {
	format: \\"jpg\\",
	type: \\"image/jpeg\\",
	name: \\"jpg1280\\",
	width: 1280,
	height: 640,
	src: __webpack_public_path__ + \\"asset.jpg\\"
}, {
	format: \\"jpg\\",
	type: \\"image/jpeg\\",
	name: \\"jpg720\\",
	width: 720,
	height: 360,
	src: __webpack_public_path__ + \\"asset.jpg\\"
}, {
	format: \\"jpg\\",
	type: \\"image/jpeg\\",
	name: \\"jpg560\\",
	width: 560,
	height: 280,
	src: __webpack_public_path__ + \\"asset.jpg\\"
}, {
	format: \\"jpg\\",
	type: \\"image/jpeg\\",
	name: \\"jpg320\\",
	width: 320,
	height: 160,
	src: __webpack_public_path__ + \\"asset.jpg\\"
}];

export var names = {
	\\"webp3000\\": __webpack_public_path__ + \\"asset.webp\\",
	\\"webp3200\\": __webpack_public_path__ + \\"asset.webp\\",
	\\"webp1920\\": __webpack_public_path__ + \\"asset.webp\\",
	\\"webp1280\\": __webpack_public_path__ + \\"asset.webp\\",
	\\"webp720\\": __webpack_public_path__ + \\"asset.webp\\",
	\\"webp560\\": __webpack_public_path__ + \\"asset.webp\\",
	\\"webp320\\": __webpack_public_path__ + \\"asset.webp\\",
	\\"jpg3000\\": __webpack_public_path__ + \\"asset.jpg\\",
	\\"jpg3200\\": __webpack_public_path__ + \\"asset.jpg\\",
	\\"jpg1920\\": __webpack_public_path__ + \\"asset.jpg\\",
	\\"jpg1280\\": __webpack_public_path__ + \\"asset.jpg\\",
	\\"jpg720\\": __webpack_public_path__ + \\"asset.jpg\\",
	\\"jpg560\\": __webpack_public_path__ + \\"asset.jpg\\",
	\\"jpg320\\": __webpack_public_path__ + \\"asset.jpg\\"
};
"
`;
